torch/csrc/distributed/c10d/ProcessGroupGloo.cpp

```diff
@@ -1923,29 +1923,68 @@ class AsyncAllgatherCUDAWork : public AsyncAllgatherWork {
   std::vector<c10::Event> outputEvents;
 };
 
+// A work that takes an lambda on construction and calls it on wait.
+// It is useful for add a continuation to another work, and/or
+// composing multiple works together.
+class LambdaWork : public Work {
+ public:
+  LambdaWork(std::function<void(void)> fn) : fn_(fn) {}
+
+  bool wait(std::chrono::milliseconds /* unused */) override {
+    fn_();
+    return true;
+  }
+
+ private:
+  std::function<void(void)> fn_;
+};
+
 } // namespace
 
 c10::intrusive_ptr<Work> ProcessGroupGloo::_reduce_scatter_base(
     at::Tensor& outputTensor,
     at::Tensor& inputTensor,
     const ReduceScatterOptions& opts) {
-  if (opts.asyncOp) {
-    throw std::runtime_error(
-        "Gloo reduce_scatter_base fallback not supported with async_op=True");
-  }
-  // first all reduce input
-  // TODO we can probably make this fully async by chaining the allreduce
-  // future.
-  std::vector<at::Tensor> inputs = {inputTensor};
-  std::vector<at::Tensor> outputs = {outputTensor};
-  allreduce(inputs)->wait();
-  std::vector<std::vector<at::Tensor>> inputTensors;
-  if (getRank() == 0) {
-    auto chunkedInputs = at::chunk(inputTensor, this->getSize(), 0);
-    inputTensors = {chunkedInputs};
-  }
-  auto scatterWork = this->scatter(outputs, inputTensors);
-  return scatterWork;
+  std::vector<at::Tensor> outputTensors = {outputTensor};
+  std::vector<at::Tensor> inputTensors = {inputTensor};
+  return reduce_scatter_tensor_coalesced(outputTensors, inputTensors, opts);
+}
+
+c10::intrusive_ptr<Work> ProcessGroupGloo::reduce_scatter_tensor_coalesced(
+    std::vector<at::Tensor>& outputTensors,
+    std::vector<at::Tensor>& inputTensors,
+    const ReduceScatterOptions& opts) {
+  if (outputTensors.size() != inputTensors.size()) {
+    TORCH_CHECK(
+        false, "requires input/output tensor lists to have the same length");
+  }
+  const auto rank = getRank();
+  const auto worldSize = getSize();
+  std::vector<at::Tensor> buffers;
+  for (const auto i : c10::irange(inputTensors.size())) {
+    auto inputShape = inputTensors[i].sizes().vec();
+    auto outputShape = outputTensors[i].sizes().vec();
+    TORCH_CHECK_EQ(outputTensors[i].dtype(), inputTensors[i].dtype());
+    TORCH_CHECK_EQ(outputShape[0] * worldSize, inputShape[0]);
+    for (size_t i = 1; i < outputShape.size(); ++i) {
+      TORCH_CHECK_EQ(outputShape[i], inputShape[i]);
+    }
+    buffers.push_back(inputTensors[i].clone());
```

Comment:
noob question about Tensor::clone, if the data is on GPU mem, will this clone be a deep copy or shallow copy (only copy pointer)?

Reply:
This result in a deepcopy, which is intentional since reduce-scatter shouldn't be modifying the input.